### PR TITLE
Update testing.parameterize using the latest version from Chainer

### PR DESCRIPTION
**Pull request by grlee77**
This PR replaces #2614 based on feedback from @kmaehashi 

This PR is mostly just a copy/paste of updated code from Chainer. The only changes are:

1.) basic renaming chainer->cupy where needed
2.) one additional commit, 8b4d900, that uses name attributes in the test IDs as formerly proposed in #2614 

Without the final commit, the full `__repr__` of classes, functions or modules gets printed instead which can be hard to read. Here is a specific case from `tests/cupy_tests/testing/test_array.py` **without** the final commit here:
``` 
tests/cupy_tests/testing_tests/test_array.py::TestEqualityAssertion_param_0_{array_module_x=<module 'numpy' from '/home/lee8rx/miniconda3_ne...r/lib/python3.7/site-packages/numpy/__init__.py'>, array_module_y=<module 'numpy' from '/home/lee8rx/miniconda3_ne...r/lib/python3.7/site-packages/numpy/__init__.py'>, assertion='assert_allclose'}.test_equality ✓1% ▏         
```

whereas with the final commit this becomes:
```
 tests/cupy_tests/testing_tests/test_array.py::TestEqualityAssertion_param_0_{array_module_x=numpy, array_module_y=numpy, assertion='assert_allclose'}.test_equality ✓                                                             1% ▎         
 ```

If that last change seems favorable, I can make a corresponding PR to chainer as well. It seems much easier to read to me, but perhaps there are other cases I haven't considered.

**One other issue to consider:**
There were two Chainer environment variables referenced in `conftest.py` that I just renamed via CHAINER->CUPY, but I'm not sure if either are needed/wanted here. Please indicate your preference regarding these. If we are keeping them, I can update the documentation accordingly.

### CODE REVIEW

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
As for `conftest.py`, how about removing the pairwise parameterization feature from this PR? It's because it is intended for reducing time for CI and is not related to the original motivation in #2614, as well as we would be able to safely remove its code from changes here.

**Comment by takagi** [comment]
I agree that 8b4d900 makes sense.

**Comment by grlee77** [comment]
> As for conftest.py, how about removing the pairwise parameterization feature from this PR?

Okay, I have now removed that file and the "pairwise" variant from the code.

**Comment by takagi** [comment]
pfnCI, test this please.

File: cupy/testing/parameterized.py [hunk 1]
```diff
@@ -32,68 +88,113 @@ def new_v(self, *args, **kwargs):
...
 
 
 def product_dict(*parameters):
+    # TODO(niboshi): Add documentation
     return [
         {k: v for dic in dicts for k, v in six.iteritems(dic)}
         for dicts in itertools.product(*parameters)]
+
+
+def _nd_indices_to_cover_each_2d(
```
Comment (takagi) [inline cupy/testing/parameterized.py]: I suppose that this function can be also removed.

Comment (grlee77) [reply cupy/testing/parameterized.py]: yes, thanks. I have updated the last commit to remove it

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
Thanks! I will merge the changes after CI pass.

**Pull request merged**